### ISSUE TITLE
Does the walker run reproduce correctly?

### ISSUE THREAD

**Issue by letusfly85** [yusukeurakami/dreamer-pytorch#4]
Hi, thank you for the cool repository!

I tried several tasks `walker walk`, `cheetah run`. They seem to work fine.

But when I run `walker run`, the episode_reward cannot achieve around 700.
Is there any problem...? 🤔 

![スクリーンショット 2020-09-12 10 35 27](https://user-images.githubusercontent.com/1466545/92984257-e8e07880-f4e3-11ea-886a-8a99a81fb8f5.png)

The original paper seems to say `walker run` will achieve around 700 with 1M steps, saying on the following page 7, Figure7.

https://arxiv.org/pdf/1912.01603.pdf

Thank you.

**Comment by yusukeurakami** [comment]
I know it is too late to comment on this issue but could you tell me the hyper parameters you tried this experiment?

**Comment by coderlemon17** [comment]
@yusukeurakami Hi, I also find that the results for `walker-run` are weird, I ran the experiment with 5 different seeds, and here's what I got:
![image](https://user-images.githubusercontent.com/39124095/223052473-da7b849d-410a-4a05-971b-0d731021303f.png)

And the hyperparameters I use are:
<details>
<summary>Hyperparameters</summary>

```
action_noise: 0.3
action_repeat: 2
actor_lr: 8.0e-05
adam_epsilon: 1.0e-07
algo: dreamer
batch_size: 50
belief_size: 200
bit_depth: 5
candidates: 1000
checkpoint_interval: 50
chunk_size: 50
cnn_activation_function: relu
collect_interval: 100
comment: ''
config: dm_control/dreamer/walker-run.yaml
dense_activation_function: elu
device: cuda:3
embedding_size: 1024
env: walker-run
episodes: 1000
exp_ckpt: ''
experience_size: 1000000
free_nats: 3
gamma: 0.99
global_kl_beta: 0.0
grad_clip_norm: 100.0
hidden_size: 200
id: dreamer
lambda_: 0.95
max_episode_length: 1000
model_ckpt: ''
model_lr: 0.001
model_lr_schedule: 0
optimisation_iters: 10
overshooting_distance: 50
overshooting_kl_beta: 0.0
overshooting_reward_scale: 0.0
planning_horizon: 15
render: false
save_experience_buffer: false
seed: 0
seed_episodes: 5
state_size: 30
symbolic_env: false
test: false
test_episodes: 10
test_interval: 10
top_candidates: 100
torch_deterministic: true
value_lr: 8.0e-05
worldmodel_LogProbLoss: false

```
</details>

**Comment by yingchengyang** [comment]
The same question. Hoping for your reply. Thanks.

**Comment by sumwailiu** [comment]
I think the root cause is that there are some minor mistakes in the reward computation (inspired by the Issue https://github.com/yusukeurakami/dreamer-pytorch/issues/5). After fixing them (refers to the pull request https://github.com/yusukeurakami/dreamer-pytorch/pull/12), I found walker-run task is reproduced correctly.

**Comment by coderlemon17** [comment]
> I think the root cause is that there are some minor mistakes in the reward computation (inspired by the Issue #5). After fixing them (refers to the pull request #12), I found walker-run task is reproduced correctly.

Hi, thanks for the explanation. However, I'm a little confused about the fix. 
Assume the sequence is $(s_1, a_1, r_1, \cdots)$. To my understanding, you are trying to predict $r_t$ with $(s_t, h_t)$, i.e. $(s_{\leq t}, a_{<t})$. However, you need $(s_{\leq t+1}, a_{<t+1})$ to predict $r_t$, is that correct?

**Comment by sumwailiu** [comment]
> Hi, thanks for the explanation. However, I'm a little confused about the fix. Assume the sequence is ( s 1 , a 1 , r 1 , ⋯ ) . To my understanding, you are trying to predict r t with ( s t , h t ) , i.e. ( s ≤ t , a < t ) . However, you need ( s ≤ t + 1 , a < t + 1 ) to predict r t , is that correct?

It is correct that I try to predict r<sub>t</sub> with (s<sub>t</sub>, h<sub>t</sub>), and I didn't use (s<sub>t+1</sub>, a<sub>t</sub>) to predict  r<sub>t</sub>. 

Acutally, the origin implementation of dreamer-torch (i.e., `reward_loss = -reward_dist.log_prob(rewards[:-1]).mean(dim=(0, 1))`) means that it tries to predict r<sub>t-1</sub> with (s<sub>t</sub>, h<sub>t</sub>). In other words, it needs (s≤t+1, a<t+1) to predict r<sub>t</sub>.

**Comment by yusukeurakami** [comment]
Thanks @sumwailiu for your PR
Let's me test on my side with your PR. I need to remember the details first. I forgot how its been implemented on the paper and original repo

**Comment by sumwailiu** [comment]
@coderlemon17 @yusukeurakami There are some mistakes in the first version of the problem fix (https://github.com/yusukeurakami/dreamer-pytorch/pull/12), where the reward computation is indeed correct. Please refer to the second version (https://github.com/yusukeurakami/dreamer-pytorch/pull/13).
